src/lua/backtrace.h

```diff
@@ -0,0 +1,110 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2022, Tarantool AUTHORS, please see AUTHORS file.
+ */
+
+#pragma once
+
+#include "trivia/config.h"
+
+#ifdef ENABLE_BACKTRACE
+#include "core/backtrace.h"
+#include "core/say.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+enum {
+	/* +2 for `backtrace_collect` and `backtrace_collect_current_stack` */
+	/* Maximal length of a Lua function  name. */
+	LUA_BACKTRACE_LUA_PROC_NAME_LEN_MAX = 64,
+	/* Maximal length of a Lua source file name. */
+	LUA_BACKTRACE_LUA_SRC_NAME_LEN_MAX = 64,
+	/* Maximal number of Lua frames collected. */
+	LUA_BACKTRACE_FRAME_COUNT_MAX = 8,
+};
+
+/*
+ * Lua frame information.
+ */
+struct lua_frame {
+	/* Ordinal number of frame. */
+	int no;
+	/* Lua function name. */
+	char proc_name[LUA_BACKTRACE_LUA_PROC_NAME_LEN_MAX];
+	/* Lua source file name. */
+	char src_name[LUA_BACKTRACE_LUA_SRC_NAME_LEN_MAX];
+	/* Line number inside Lua function. */
+	int line_no;
+};
+
+/*
+ * Collection of Core and Lua frames.
+ */
+struct lua_backtrace {
+	/* Core backtrace information. */
+	struct core_backtrace core_bt;
+	/* Ordinal number of first Lua frame in backtrace. */
+	int first_lua_frame_no;
+	/* Number of Lua frames collected. */
+	int frame_count;
+	/* Array of Lua frame data. */
+	struct lua_frame frames[LUA_BACKTRACE_FRAME_COUNT_MAX];
```

Comment:
I may be wrong, but AFAIK there may be several Lua stacks within a stack: a C function may call a Lua function (see [`luaL_loadstring`](https://www.lua.org/manual/5.1/manual.html#luaL_loadstring)), which in turn may call another C function, and so on. AFAICS this backtrace implementation wouldn't work in this case.

Reply:
To illustrate the problem, I ran the following script before and after the first commit of this PR:

<details><summary>Test script</summary>

```lua
local fiber = require('fiber')
local net = require('net.box')
local yaml = require('yaml')

box.cfg{listen = 3301, log_level = 'warn'}

local c = net.connect(box.cfg.listen)
local info
c:on_disconnect(function()
    info = fiber.self():info()
end)
c:close()

print(yaml.encode(info))
os.exit(0)
```

</details>

BTW the first commit doesn't build. To fix it I had to apply the following patch:

<details><summary>Build fix</summary>

```diff
diff --git a/src/lib/core/backtrace.cc b/src/lib/core/backtrace.cc
index 10498e25216a..1e28cdbc6048 100644
--- a/src/lib/core/backtrace.cc
+++ b/src/lib/core/backtrace.cc
@@ -43,6 +43,7 @@ __attribute__ ((__force_align_arg_pointer__))
 static NOINLINE void *
 collect_current_stack(struct core_backtrace *bt, void *stack)
 {
+#ifndef __APPLE__
 	bt->frame_count = unw_backtrace((void **)bt->frames,
 					CORE_BACKTRACE_FRAME_COUNT_MAX);
 #else /* __APPLE__ */
```

</details>

Here are the results:

<details><summary>Before</summary>

```
---
csw: 26
backtrace:
- C: '#0  0x5616c8366603 in lbox_fiber_object_info+130'
- L: info in =[C] at line -1
- L: func in @../test.lua at line 10
- L: run in @builtin/internal.trigger.lua at line 61
- L: (unnamed) in @builtin/box/net_box.lua at line 289
- C: '#1  0x5616c83d89bf in lj_BC_FUNCC+52'
- C: '#2  0x5616c83db419 in jit_secure_call+445'
- C: '#3  0x5616c83e3240 in lua_call+199'
- C: '#4  0x5616c8351be5 in netbox_transport_on_state_change+229'
- C: '#5  0x5616c8353338 in luaT_netbox_transport_stop+302'
- L: stop in =[C] at line -1
- L: close in @builtin/box/net_box.lua at line 665
- L: (unnamed) in @../test.lua at line 12
- C: '#6  0x5616c83d89bf in lj_BC_FUNCC+52'
- C: '#7  0x5616c83e3597 in lua_pcall+849'
- C: '#8  0x5616c836c588 in luaT_call+45'
- C: '#9  0x5616c8364210 in lua_main+189'
- C: '#10 0x5616c836475d in run_script_f+1332'
- C: '#11 0x5616c8140977 in fiber_cxx_invoke(int (*)(__va_list_tag*), __va_list_tag*)+34'
- C: '#12 0x5616c8393ad6 in fiber_loop+136'
- C: '#13 0x5616c870944a in coro_init+88'
memory:
  total: 516480
  used: 0
time: 0
name: test.lua
fid: 103
...
```

</details>

<details><summary>After</summary>

```
---
csw: 26
backtrace:
- C: '#1  0x561df602256b in lbox_fiber_statof_map+808'
- C: '#2  0x561df60230f5 in lbox_fiber_object_info+130'
- L: '#3  info in =[C]:-1'
- L: '#4  func in @../test.lua:10'
- L: '#5  run in @builtin/internal.trigger.lua:61'
- L: '#6  (unnamed) in @builtin/box/net_box.lua:289'
- C: '#7  0x561df60959df in lj_BC_FUNCC+52'
- C: '#8  0x561df6098439 in jit_secure_call+445'
- C: '#9  0x561df60a0260 in lua_call+199'
- C: '#10 0x561df600ebf5 in netbox_transport_on_state_change+229'
- C: '#11 0x561df6010348 in luaT_netbox_transport_stop+302'
- C: '#12 0x561df60959df in lj_BC_FUNCC+52'
- C: '#13 0x561df60a05b7 in lua_pcall+849'
- C: '#14 0x561df602907a in luaT_call+45'
- C: '#15 0x561df602119f in lua_main+189'
- C: '#16 0x561df60216ec in run_script_f+1332'
- C: '#17 0x561df5dfd987 in fiber_cxx_invoke(int (*)(__va_list_tag*), __va_list_tag*)+34'
- C: '#18 0x561df60515b6 in fiber_loop+136'
- C: '#19 0x561df63c648a in coro_init+88'
memory:
  total: 516480
  used: 0
time: 0
name: test.lua
fid: 103
...
```

</details>

It doesn't look like mere refactoring.

Reply:
Fixed: 
* redesigned `struct lua_frame`: now it is union plus a enum for reflection — I guess its the best way to express the derived relationship between it and `struct core_frame`;
* fixed up refactoring commit (it contained a line from the next commit).

Now your script shows the same results.